lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,774 @@
+    end
+  end
+
+  defp validate_start({m, f, args}) when is_atom(m) and is_atom(f) and is_list(args), do: :ok
+  defp validate_start(mfa), do: {:invalid_mfa, mfa}
+
+  defp validate_type(type) when type in [:supervisor, :worker], do: :ok
+  defp validate_type(type), do: {:invalid_child_type, type}
+
+  defp validate_restart(restart) when restart in [:permanent, :temporary, :transient], do: :ok
+  defp validate_restart(restart), do: {:invalid_restart_type, restart}
+
+  defp validate_shutdown(shutdown) when is_integer(shutdown) and shutdown > 0, do: :ok
+  defp validate_shutdown(shutdown) when shutdown in [:infinity, :brutal_kill], do: :ok
+  defp validate_shutdown(shutdown), do: {:invalid_shutdown, shutdown}
+
+  defp validate_modules(:dynamic), do: :ok
+
+  defp validate_modules(mods) do
+    if is_list(mods) and Enum.all?(mods, &is_atom/1) do
+      :ok
+    else
+      {:invalid_modules, mods}
+    end
+  end
+
+  @doc """
+  TODO
+  """
+  @spec terminate_child(Supervisor.supervisor(), pid) :: :ok | {:error, :not_found}
+  def terminate_child(supervisor, pid) when is_pid(pid) do
+    call(supervisor, {:terminate_child, pid})
+  end
+
+  @doc """
+  Returns a list with information about all children.
+
+  Note that calling this function when supervising a large number
+  of children under low memory conditions can cause an out of memory
+  exception.
+
+  This function returns a list of tuples containing:
+
+    * `id` - as defined in the child specification but is always
+      set to `:undefined` for dynamic supervisors
+
+    * `child` - the pid of the corresponding child process or the
+      atom `:restarting` if the process is about to be restarted
+
+    * `type` - `:worker` or `:supervisor` as defined in the child
+      specification
+
+    * `modules` - as defined in the child specification
+
+  """
+  @spec which_children(Supervisor.supervisor()) :: [
+          {:undefined, pid | :restarting, :worker | :supervisor, :supervisor.modules()}
+        ]
+  def which_children(supervisor) do
+    call(supervisor, :which_children)
+  end
+
+  @doc """
+  Returns a map containing count values for the supervisor.
+
+  The map contains the following keys:
+
+    * `:specs` - always 1 as dynamic supervisors have a single specification
+
+    * `:active` - the count of all actively running child processes managed by
+      this supervisor
+
+    * `:supervisors` - the count of all supervisors whether or not the child
+      process is still alive
+
+    * `:workers` - the count of all workers, whether or not the child process
+      is still alive
+
+  """
+  @spec count_children(Supervisor.supervisor()) :: %{
+          specs: non_neg_integer,
+          active: non_neg_integer,
+          supervisors: non_neg_integer,
+          workers: non_neg_integer
+        }
+  def count_children(supervisor) do
+    call(supervisor, :count_children)
+  end
+
+  @doc """
+  TODO
+  """
+  def init(options) when is_list(options) do
```

Comment:
Isn't it weird to have `init/1` called both from outside the supervisor and from the inside?

Reply:
It is but I am out of ideas since we want to mirror `Supervisor.init`.

Reply:
I guess one option would be to split the public API module and GenServer callback module, but that adds some mental overhead to implementation.